streams/src/main/java/org/apache/kafka/streams/kstream/internals/WindowedKStreamImpl.java

```diff
@@ -0,0 +1,143 @@
+import org.apache.kafka.streams.state.Stores;
+import org.apache.kafka.streams.state.WindowStore;
+
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.apache.kafka.streams.kstream.internals.KGroupedStreamImpl.AGGREGATE_NAME;
+import static org.apache.kafka.streams.kstream.internals.KGroupedStreamImpl.REDUCE_NAME;
+
+public class WindowedKStreamImpl<K, V, W extends Window> extends AbstractStream<K> implements WindowedKStream<K, V> {
+
+    private final Windows<W> windows;
+    private final boolean repartitionRequired;
+    private final Serde<K> keySerde;
+    private final Serde<V> valSerde;
+
+    WindowedKStreamImpl(final Windows<W> windows,
+                        final InternalStreamsBuilder builder,
+                        final Set<String> sourceNodes,
+                        final String name,
+                        final Serde<K> keySerde,
+                        final Serde<V> valSerde,
+                        final boolean repartitionRequired) {
+        super(builder, name, sourceNodes);
+        Objects.requireNonNull(windows, "windows can't be null");
+        this.valSerde = valSerde;
+        this.keySerde = keySerde;
+        this.repartitionRequired = repartitionRequired;
+        this.windows = windows;
+    }
+
+    @Override
+    public KTable<Windowed<K>, Long> count() {
+        return aggregate(
+                new Initializer<Long>() {
+                    @Override
+                    public Long apply() {
+                        return 0L;
+                    }
+                }, new Aggregator<K, V, Long>() {
+                    @Override
+                    public Long apply(K aggKey, V value, Long aggregate) {
+                        return aggregate + 1;
+                    }
+                },
+                Serdes.Long());
+    }
+
+    @Override
+    public <VR> KTable<Windowed<K>, VR> aggregate(final Initializer<VR> initializer,
+                                                  final Aggregator<? super K, ? super V, VR> aggregator,
+                                                  final Serde<VR> aggValueSerde) {
+        Objects.requireNonNull(initializer, "initializer can't be null");
+        Objects.requireNonNull(aggregator, "aggregator can't be null");
+        final String aggFunctionName = builder.newName(AGGREGATE_NAME);
+        final String storeName = builder.newStoreName(AGGREGATE_NAME);
+        return doAggregate(aggValueSerde,
+                           aggFunctionName,
+                           storeName,
+                           new KStreamWindowAggregate<>(windows, storeName, initializer, aggregator));
+    }
+
+    @Override
+    public KTable<Windowed<K>, V> reduce(final Reducer<V> reducer) {
+        Objects.requireNonNull(reducer, "reducer can't be null");
+        final String storeName = builder.newStoreName(REDUCE_NAME);
+        return doAggregate(valSerde,
+                           REDUCE_NAME,
```

Comment:
Is this correct, to directly using the `REDUCE_NAME`?

Reply:
Thanks, missing the `builder.newName(..)`